Feature/remotes

alt-graph opened the pull request:
This MR adds functionality for dealing with git remote repositories and for pushing refs from a local repository to a remote one. It disables the previous, incomplete implementation of `pull()` and related functionality.

The core is a new `Remote` class that encapsulates functionality for dealing with a git remote repository. The new `GitRepository` member functions `add_remote()`, `get_remote()`, and `list_remotes()` make use of it. In addition, the previous implementation of `push()` is made more functional by allowing the user to specify both a remote and a refspec (saying what should be pushed where).

As a drive-by cleanup, all files generated by unit tests are now collected under unit_test_files/.

alt-graph committed:
Add Remote class

[why]
As a lightweight abstraction over the C git_remote struct handed out by
libgit2, the Remote class mainly gives easy access to the remote name
and URL.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Disable previous remote functionality

[why]
GitRepository assumed that there could be only a single associated git
remote, but git allows an arbitrary number of them. The functionality
needs to be extended to deal with that.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
GitRepository: Add add_remote(), get_remote(), list_remotes()

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Remove unnecessary namespace qualifiers

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Move all files produced by tests under unit_test_files/

[why]
To keep them in one place, and to be able to remove them with a single
command in the unit test main().

[how]
Introduce a function unit_test_folder() in main.h/cc that returns the
name of the folder, "unit_test_files", and use it for constructing file
names in the tests.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
GitRepository: Implement push() with remote and refspec

[why]
The new implementation allows specifying the remote to push to and which
local object(s) to push to which remote object(s) via a git refspec.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Move credentials_callback() into its own source file

[why]
The credentials callback function is central for libgit2 whenever remote
repositories need to be connected (and the protocol requires usernames,
passwords, or something similar). It will be needed from both the
GitRepository and Remote classes.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Remote: Implement list_references()

[why]
To retrieve a list of references from a remote repository à la "git
ls-remote".

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Extend unit test for push()

This commit adds a verification that push() actually updates the remote
repository.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph commented:
@wyrnat, @Finii, @MissPeppermintPatty: Does anyone have time to review this?

Finii commented:
> @wyrnat, @Finii, @MissPeppermintPatty: Does anyone have time to review this?

I can do sometime in the next 48h

Finii commented on src/credentials_callback.cc:
```diff
@@ -0,0 +1,38 @@
+    const char* /*username_from_url*/, unsigned int /*allowed_types*/, void* /*payload*/)
+{
+    // Dummy implementation: Always return the same username and password
+    return git_cred_userpass_plaintext_new(out_credentials, "user", "pwd");
```
:thinking: 

4 years ago they changed `cred` to `credential`.
The API documentation lists (only) the long forms.

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/7f9c0b7f-e639-48e5-854b-15329352b9d9)

Here the commit over there:

https://github.com/libgit2/libgit2/commit/3f54ba8b61869f42b2bbd1a60091a0be640bc8fc

## Package situation

Ubuntu 20.04 has the ancient libgit, while Ubuntu 22.04 has libgit 1.1.0 - **so this PR will not build on Ubuntu 22.04**

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/d71068c2-9152-4d30-990d-0d4667ffac01)

I'm not sure why it builds in the CI... compare `doocsdev22` below

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/504f00b5-aeb8-48cd-83af-6a4f9612d785)

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/0fc04988-24e9-4bf7-8762-65ffaf3a80bc)
_Error message is misleading, but still look on green suggested solution_

Finii replied:
Hmm, I remember that I had some `meson.build` that checked the `gitlib2` version and
* Output "Ancient libgit2 version found" (or similar)
* Set some macro that code could use in `#ifdef`
* Used that macro for some library function that changed behavior with v1.0.0

But I can not find it and do not see it anymore :thinking:

Finii replied:
Ah, here:

```python
## libgit2 has obviously a change in the globbing
local_compiler_args = []
if dependency('libgit2').version().version_compare('<1')
    message('Ancient libgit2 version detected')
    local_compiler_args += [ '-DANCIENT_LIBGIT2' ]
endif
```

The comparison is obviously wrong because the change happened already with 0.99.0, but that is not released as package because it is a preview version.

https://github.com/taskolib/taskolib/commit/afb384be8173e5b4e1a5338df68af8cbb32c49df

And ... why does `taskolib` need to know about the libgit details, that should be encapsulated here?

Finii replied:
Oh, ...
at least the Ubuntu package also has the old names as symbols, that is the reason it compiles on the CI. My own runs had errors and suggested were by chance the longer new names.

But all are there:

```console
$ nm -DC /usr/lib/x86_64-linux-gnu/libgit2.so.1.1.0 | grep git_cred 
00000000000a7340 T git_cred_default_new
00000000000a7120 T git_credential_default_new
00000000000a7170 T git_credential_free
00000000000a6e00 T git_credential_get_username
00000000000a6df0 T git_credential_has_username
00000000000a7040 T git_credential_ssh_custom_new
00000000000a6f10 T git_credential_ssh_interactive_new
00000000000a6fb0 T git_credential_ssh_key_from_agent
00000000000a6f00 T git_credential_ssh_key_memory_new
00000000000a6ef0 T git_credential_ssh_key_new
00000000000ad1f0 T git_credential_username_new
00000000000ad280 T git_credential_userpass
00000000000a6e40 T git_credential_userpass_plaintext_new
00000000000a73a0 T git_cred_free
00000000000a7320 T git_cred_get_username
00000000000a7310 T git_cred_has_username
00000000000a7390 T git_cred_ssh_custom_new
00000000000a7370 T git_cred_ssh_interactive_new
00000000000a7380 T git_cred_ssh_key_from_agent
00000000000a7360 T git_cred_ssh_key_memory_new
00000000000a7350 T git_cred_ssh_key_new
00000000000ad270 T git_cred_username_new
00000000000ad2e0 T git_cred_userpass
00000000000a7330 T git_cred_userpass_plaintext_new
```

We still might need to know if we can utilize the new return values or whatever, and maybe put the macro in HERE instead of into taskolib.

alt-graph replied:
I have imported the check into the meson.build file, but commented it out as we apparently do not need it (yet)? See commit 502b51bbc20f4ae751afbc7d1a5c585a4b95313f, we can also drop it if preferred. I am going to remove that part from Taskolib.

alt-graph replied:
>  I am going to remove that part from Taskolib.

Actually I cannot remove it because the test for ANCIENT_LIBGIT2 is used inside Taskolib to determine if a glob should end in "/*" or not. So if the test is needed in Taskolib, it should also be done in Taskolib.

Finii replied:
Well I meant ... should not `libgit4cpp` represent one globbing way, and IT adapts the glob to ancient libgit if needed?
So the user does not nee to know that the globbing might change with underlying version.

This is just an idea. The macro is kind of suboptimal, because one needs to know that it is needed in your own code.
Maybe I will try to come up with a PR that unifies the globbing stuff; but I guess before that I would need to examine when it has been introduced into libgit and why and what else etc pp. ;-)

Finii commented:
I'm quite opposed to this commit

https://github.com/taskolib/libgit4cpp/pull/18/commits/978cb8c743e312003c3a4c2c84b9e94c6596f035  `Remove unnecessary namespace qualifiers`

which does (almost only) changes regarding `Error` like:

```diff
-throw git::Error{ "Git init failed" };
+throw Error{ "Git init failed" };
```

We have an `Error` class in all the libraries, which is imho already error prone, but at least we give the namespace to make clear what it is. In the actual server you might juggle `doocs::Error`, `git::Error`, and `task::Error` and I find that too confusing and hostile for grepping.
It might be superfluous here (because this is the lowest in the chain), but in principle I would keep the namespace specifically for all the different Error objects, because they are too similar jet dissimilar - dropping the namespace raises instead of lowers the mental burden in my opinion.

Finii commented on src/credentials_callback.h:
```diff
@@ -0,0 +1,44 @@
+#define LIBGIT4CPP_CREDENTIALS_CALLBACK_H_
+
+namespace git {
+extern "C" {
```
I'm not sure this is a good idea (extern C within a namespace).

From C++ and when we compile the function name is `git::credentials_callback()`, but of course the symbol later on will be just `credentials_callback()` because it should be callable/linkable with C.

When some person uses libgit4cpp and creates a `my_very_own::credentials_callback` in a similar way we will end up with two identically named symbols.

If we do not put it in the namespace the (C++) compiler might warn earlier on.

I find the following explanation makes it most obvious.
The code is kind of misleading imho.

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/0c0ad5b0-7b03-4f86-84ca-b3647200efda)

alt-graph replied:
> The code is kind of misleading imho.

I agree, the namespace does not make any sense. I'll rename the function to something ugly like `libgit4cpp_credentials_callback()`.

alt-graph replied:
I had another idea: The latest commit hides the C function away with internal linkage and implements a nice public `CredentialsCallback get_dummy_credentials_callback()` that returns a function pointer to it.

Finii commented on include/libgit4cpp/GitRepository.h:
```diff
@@ -170,15 +171,49 @@ class GitRepository
+    gul14::optional<Remote> get_remote(const std::string& remote_name) const;
 
+    /**
+     * Return a list of all configured git remotes.
```
I think what the function description says does make more sense than what the function actually does.

* _Description:_ Return a list of all remotes
* _Actually does:_ Return `Remote` objects for all remotes

Usually, I guess, you are only interested in one remote (at a time) and throw all the other `Remote` objects away.

alt-graph replied:
Does that distinction matter? Yes, we spend a tiny amount of extra work to retrieve the URLs as well and to make them easily available, but... why not? Sure, we can return a list of remote names and make the user retrieve the repository pointer, and then they can retrieve the URL if they like. In that case we do not need the `Remote` class at all. It is just an attempt to raise the level of abstraction.

Finii replied:
At least it should be called `get_all_remotes()` or similar.

And the 'tiny amount of extra work' is what makes you use the less readable `gul::cat()` instead of much more readable `{fmt}` or similar. "But cat saves a tiny amount of work", yes right, which is of course very important when an error message is constructed which is not even in the vicinity of a hot path.

I stumbled about a related thing with `SequenceManager::list_sequences()`. Once I put a debug message in the inside and was astonished how often it is called in an ordinary server.
Example:
* Want to remove a Sequence: `remove_sequence(UniqueId)`
* Calls `list_sequences()`
* Checks every single directory entry if it is a Sequence and constructs many many strings
* Now `find_sequence_on_disk(unique_id, sequences)`
* Walk trough the vector `list_sequences()` returned and search for the one with the right UniqueID to find the path
* Now that Sequence is removed

This does a lot extra work which is never used. Instead a `find_sequence_on_disk(unique_id)` would be more to the point because it could search much more direct. Or a `list_sequences(glob)`.

> In that case we do not need the Remote class at all

Yes. The need for that class is indeed a question. The previous approach with a `remote_` data member in the `GitSequence` was not bad either. The only limitation was that it was not a vector but only one single remote, which can be limiting (but then, that could also be changed without a new class `Remote`).

Anyhow, a `Remote` object by itself is not too helpful as it does not even remember which repo it belonged to. If I want to handle multiple repos and their remotes that becomes a hurdle. I as user need to keep the association myself somewhere. I would not need to do that if the remote is an integral part of the repository object.

Ok.

I try to see where and how `list_remotes()` would be used.
If I want to list the existing remotes I would just need the names as string.
If I want to modify one specific remote I already have the target-remote name somehow and can select that remote directly.
If I want to push to a specific remote I have the target-remote name already; checking if the name exists does not need all remotes as `Remote`, this can be decided on the `push()` call specifically for that name (compare Sequence example above), and/or I can live with a vector of names.

And then, why do we have the members AT ALL?

```c++
class Remote
{
    /// Return a non-owning pointer to the underlying git remote object.
    git_remote* get() const { return remote_.get(); }

    const std::string& get_name() const { return name_; }
    const std::string& get_url() const { return url_; }

private:
    std::string name_;
    std::string url_;
    LibGitRemote remote_{ nullptr, git_remote_free };
```
versus
```c++
class Remote
{
    /// Return a non-owning pointer to the underlying git remote object.
    git_remote* get() const { return remote_.get(); }

    std::string get_name() const { return git_remote_name(remote_.get()); }
    std::string get_url() const { return git_remote_url(remote_.get()); }

private:
    LibGitRemote remote_{ nullptr, git_remote_free };
```

I do not see why we need the members at all. And with just the `LibGitRemote` object as member, `Remote` is kind of ... useless? :thinking: 

When I examine where `Remote` is used, e.g. `push()` only the `LibGitRemote` member is utilized at all.

Finii replied:
Just for comparison, what other people try to put into the `Remote` object:

```c++
namespace git
{
    struct Remote
    {
        const char * url()      const;
        const char * pushurl()  const;

        struct FetchCallbacks
        {
            virtual void update_tips(char const * refname, git_oid const & a, git_oid const & b) {}
            virtual void sideband_progress(char const * str, int len) {}
            virtual void transfer_progress(git_indexer_progress const &) {}

            virtual git_credential* acquire_cred(const char * url, const char * username_from_url, unsigned int allowed_types) = 0;
        };

        void fetch(FetchCallbacks &, char const * reflog_message = nullptr);

    private:
        struct Destroy { void operator() (git_remote*) const; };

        explicit Remote(git_remote * remote) : remote_(remote) {}

        std::unique_ptr<git_remote, Destroy> remote_;
    };
}
```

https://github.com/AndreyG/libgit2cpp/blob/master/include/git2cpp/remote.h

Finii commented on include/libgit4cpp/Remote.h:
```diff
@@ -0,0 +1,82 @@
+private:
+    std::string name_;
+    std::string url_;
+    LibGitRemote remote_{ nullptr, git_remote_free };
```
We have no connection (anymore) between `Remote` and `GitRepository` anymore.
But that might come in handy.

I'm also not sure that `Remote` should be allowed to outlive the originating `GetRepository`, because
`git::Remote.get()->git_remote_owner()` return a pointer to the repository object that might or might not be living anymore.

This is of course a complex situation that has previously been solved by having the remote object a member of the `GitRepository`, so that the repo can not die before the remote.

One solution could be that the `Remote` holds a `weak_ptr` to the `GitRepository` and on any access check if the repo is still alive.

The vanishing original owner problem is for example solved in the DOOCSdzmq library with the `safe_self_` object.

<details>

Here the code snippet. The situation there is simular but different. The MacroPulseChecker holds pointers to all the properties that have registered. But the properties might have been removed in the meantime.

I hope this makes sense :-D

Showing the code here, because reliably generating the shared pointer without much manual work can be confusing. (I mean in the constructor, where we are not allowed to access `this` yet. Also can not remember why the pointer is explicitly reset to `nullptr` on destruction, but for sure there was a reason :thinking: 

```c++
class D_ZMQ  {
  private:
       static MacroPulseChecker MPchecker_;
       D_ZMQ*                   self_{this};
  protected:
       std::shared_ptr<D_ZMQ*>  safe_self_;
}

D_ZMQ::D_ZMQ (const std::string &pn, EqFct *ef, bool sender)
{
    safe_self_ = std::make_shared<D_ZMQ*>(self_);
}

D_ZMQ::~D_ZMQ (void)
{
    safe_self_ = std::make_shared<D_ZMQ*>(nullptr);
}

void D_ZMQ::register_MPchecker()
{
    MPchecker_.push_back(MPchecker_idx_, prop, safe_self_); // safe_self_ taken as weak_ptr
}

// And here we have a lot of property pointer that might or might not still exist
// Remove them if the property has been removed meanwhiles...

std::string MacroPulseChecker::get_fill_status(int bufno)
{
            if (e.prop_ptr.expired())
                this->remove(idx); // remove removed properties
}
```

I do not want to suggest this code for `Remote` or `GitRepository`, but we should have some solution.

</detail>

alt-graph replied:
> I'm also not sure that Remote should be allowed to outlive the originating GetRepository, because
git::Remote.get()->git_remote_owner() return a pointer to the repository object that might or might not be living anymore.

I see your point, but I do not think we need to address that. IMHO if users start making their own calls into `libgit2`, they are responsible for their own mess. I would see things differently if you couldn't safely call `Remote` member functions anymore once the parent repo has been destructed, but that is not the case. (... unless it is?)

Finii replied:
The `git_remote` object:

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/aecc252a-d883-4f67-8934-0159d96df39a)

That is filled directly with the same pointer value as out `GitRepository.get()` (but now non-owning), thus is invalid if the repository goes out of scope.

See https://github.com/libgit2/libgit2/blob/25e2b9d8c217e92eb70cde85bb4007a3dec0428d/src/libgit2/remote.c#L502

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/7553bd84-a88c-4e5b-8cbc-5ab122b70e5e)

That seems to be used by the `transport` that is used for example by `git_remote_ls()`, if ssh is used?
Well I did not do an extensive search.

For me it seems a `git_remote` does need the existing `git_repository` and we should make sure of that even if that is more work - is that not what C++ is for in comparison to C, that dangling pointers are avoided?

And further I also think they are responsible for their own mess, but WE should try to prevent any footguns we can; and that is doable for `git::Remote` that know if their respective `git::GitRepository` has been destructed.
If that is too much work I can implement that, did that a couple of times already for similar cases.

Finii replied:
At least we should try what comes out with asan if
* Create a `GitRepository` for an existing git repo on the filesystem that has remotes
* Crate a `Remote` from that
* Destruct the `GitRepository`
* See if `Remote::listRreferences()` is an asan problem

If yes the need is clear.
If not we can discuss if footgun prevention is needed or documentation is enough.

alt-graph replied:
I have added a test for that: 2d06bb77d2596a43acc1b96cde25a41b82070498

No problems with the address sanitizer on Ubuntu 20 LTS. I see no need for further safeguards.

Finii replied:
Great! Thanks for performing the test.
Then I believe we leave it as it is and possibly improve it later (if at all ;-> )

Finii commented on include/libgit4cpp/Remote.h:
```diff
@@ -0,0 +1,82 @@
+/**
+ * \file   Remote.h
```
Why is this `Remote` and not `GitRemote` :thinking: 

It is `GitRepository` instead of just `Repository`, and I always believe very generic names like `name` for functions or member, or `Remote` are sometimes a reason for confusion.

alt-graph replied:
I find `git::GitRepository` redundant and I would actually prefer to rename `git::GitRepository->git::Repository` instead of `git::Remote->git::GitRemote`.

Finii replied:
Then lets do that (in a next PR)

Finii commented on include/libgit4cpp/wrapper_functions.h:
```diff
@@ -120,12 +120,12 @@ LibGitRemote remote_create (git_repository* repo, const std::string& remote_name
 
 /**
- * Collects the remove connection by name.
+ * Find a remote repository by name.
```
I think both (old and new) are not to the point.

It does not find a remote repository, unless it has been added. :grimacing: 
Of course that is what you mean. But still.

The old description is not nice, but the new does not fix that either. :thinking:

alt-graph replied:
I have tried to rephrase...

Finii replied:
Very good :+1:

Finii commented on src/GitRepository.cc:
```diff
@@ -429,31 +427,82 @@ void GitRepository::reset(unsigned int nr_of_commits)
 }
 
-
+#if 0
```
I would throw out commented-out code, we can always go back.
Unless you expect that it will get revived later on.

alt-graph replied:
It will get revived in one of the next MRs. If we have `push()`, we need `pull()`, don't we?

Finii replied:
Sure.

I just looked at the diffs (well, mostly). And as I said, _unless you expect_

Finii commented on src/Remote.cc:
```diff
@@ -0,0 +1,81 @@
+        throw Error{ "Remote pointer may not be null" };
+
+    name_ = gul14::safe_string(git_remote_name(remote_.get()), 512);
+    url_ = gul14::safe_string(git_remote_url(remote_.get()), 512);
```
What is the magic of `512`?
And why is it not `500` (I will never understand why limits should be power-of-2 for no technical reason)?

I also do not understand why you expect an un-terminated string here?

`libgit` does make sure it is valid (well, at least it also calls `strlen()` on it, unlimited, before you "do the same" but more anxious.
Also other returned strings are not handed like this (e.g. from `git_remote_list()`)?

I just don't get the reason for the `safe_string` when internally by `libgit` the strings are of course used directly.

https://github.com/libgit2/libgit2/blob/25e2b9d8c217e92eb70cde85bb4007a3dec0428d/src/libgit2/remote.c#L630-L662

https://github.com/libgit2/libgit2/blob/25e2b9d8c217e92eb70cde85bb4007a3dec0428d/src/util/alloc.c#L51-L60

Example

```c++
static int canonicalize_url(git_str *out, const char *in)
{
	if (in == NULL || strlen(in) == 0) {
		git_error_set(GIT_ERROR_INVALID, "cannot set empty URL");
		return GIT_EINVALIDSPEC;
	}
```

and afterwards it is dupped and put into the member.

alt-graph replied:
> What is the magic of `512`? And why is it not `500` (I will never understand why limits should be power-of-2 for no technical reason)?

No magic. Whatever.

> I also do not understand why you expect an un-terminated string here?

I do not, but I do expect nullptr: "a pointer to the name or NULL for in-memory remotes"

> `libgit` does make sure it is valid (well, at least it also calls `strlen()` on it, unlimited, before you "do the same" but more anxious. Also other returned strings are not handed like this (e.g. from `git_remote_list()`)?

It is not necessarily valid. It might be null.

Finii replied:
But `safe_string` also also about a length limit (presumably to avoid ... what? Non-terminated string crashes?)

Then why not the much clearer
```diff
-name_ = gul14::safe_string(git_remote_name(remote_.get()), 512);
+auto n = git_remote_name(remote_.get());
+name_ = n ? n : ""; // or something? Is "" a good name for file remotes?
```
that does not break for otherwise working idiotic long names :->

alt-graph replied:
> Then why not the much clearer
> 
> ```diff
> -name_ = gul14::safe_string(git_remote_name(remote_.get()), 512);
> +auto n = git_remote_name(remote_.get());
> +name_ = n ? n : ""; // or something? Is "" a good name for file remotes?
> ```
> 
> that does not break for otherwise working idiotic long names :->

Fair enough, changed.

Finii commented on src/Remote.cc:
```diff
@@ -0,0 +1,81 @@
+    std::vector<std::string> refs;
+    refs.reserve(size);
+    for (size_t i = 0; i != size; ++i)
+        refs.emplace_back(out[i]->name);
```
No `safe_string` :+1:

alt-graph replied:
Because the libgit2 API docs do not specify nullptr as a possible result.

Finii commented on src/wrapper_functions.cc:
```diff
@@ -118,12 +118,9 @@ LibGitRemote remote_create(git_repository* repo, const std::string& remote_name,
-        remote = nullptr;
-    }
+    git_remote* remote = nullptr;
+    if (repo)
```
Silently ignore unset/empty `repo`? :thinking: 

`git_remote_lookup()` will check for a `nullptr` unless `NDEBUG` - I have no clue how it is compiled.
It will crash otherwise (`repo->_config` is called).

The other wrappers just assume a valid `repo`.

Whatever we do, we should do it everywhere (not necessarily within this PR).

alt-graph replied:
> Whatever we do, we should do it everywhere (not necessarily within this PR).

That is of course true. For the lookup especially I found this to be reasonable – no repository, no remotes. Will defer this to a later PR.

Finii commented on tests/test_GitRepository.cc:
```diff
@@ -71,7 +74,7 @@ void create_testfiles(const std::filesystem::path& name, size_t nr_files,
 } // anonymous namespace
 
-TEST_CASE("GitRepository Wrapper Test all", "[GitWrapper]")
+TEST_CASE("GitRepository Wrapper Test all", "[GitRepository]")
```
:thinking: 

```suggestion
TEST_CASE("GitRepository Test all", "[GitRepository]")
```

Additionally I would suggest to break this apart, for me there are far too many unrelated `SECTION`s, which makes debugging more hard. (Breaking apart also of course not in the PR)

alt-graph replied:
Ack.

Finii commented on tests/test_GitRepository.cc:
```diff
@@ -575,19 +578,110 @@ TEST_CASE("GitRepository add() with glob", "[GitWrapper]")
  * Tidy up) Reset remote, delete all local files
- 
+
 TEST_CASE("GitRepository Wrapper Test Remote", "[GitWrapper]")
```
`GitWrapper` :clown_face: 

Ok, this is commented out.

Maybe remove commented out stuff (see some comment above)

Finii commented on tests/test_main.cc:
```diff
@@ -22,8 +22,20 @@
+
+std::filesystem::path unit_test_folder()
+{
+    return "unit_test_files";
```
:+1: :+1: :+1:

Finii reviewed (commented):
This looks like a tremendous amount of work done!
Nice PR indeed.

I had only time to go through the changes by eye.
Will do some functional tests later, but wanted to already feed back.

Finii commented:
Ah, I forgot, `branch_remote_name` and thus
* #20 

is not directly needed anymore, but I believe the error handling introduced there should still prevail / be added. This PR adds a lot of unused return values. ;)

alt-graph committed:
Return dummy credentials callback via getter()

Instead of publishing a global function with a C interface, this commit
implements a getter that returns a function pointer to such a function
and hides the function away with internal linkage.

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Improve documentation for remote_lookup()

Requested-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph commented:
> I'm quite opposed to this commit
> 
> [978cb8c](https://github.com/taskolib/libgit4cpp/commit/978cb8c743e312003c3a4c2c84b9e94c6596f035) `Remove unnecessary namespace qualifiers`
> 
> which does (almost only) changes regarding `Error` like:
> 
> ```diff
> -throw git::Error{ "Git init failed" };
> +throw Error{ "Git init failed" };
> ```
> 
> We have an `Error` class in all the libraries, which is imho already error prone, but at least we give the namespace to make clear what it is. In the actual server you might juggle `doocs::Error`, `git::Error`, and `task::Error` and I find that too confusing and hostile for grepping. It might be superfluous here (because this is the lowest in the chain), but in principle I would keep the namespace specifically for all the different Error objects, because they are too similar jet dissimilar - dropping the namespace raises instead of lowers the mental burden in my opinion.

I understand your point, but at least for `libgib4cpp` it does not make sense to me. The project has no knowledge of Taskolib or DOOCS and there are no conflicting `Error` classes anywhere. IMO repeating the namespace qualifier just adds visual noise and makes the code harder to read.

alt-graph committed:
Remote: Change initialization of name_, url_

Proposed-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Meson: Add commented-out check for libgit2 version

[why]
libgit2 has changed its API at version 1. It will probably become
necessary to enable conditional compilation in the future, and this is
the code snippet for it.

Proposed-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Remote: Test list_references() still working if parent repo removed

The new test removes the parent repo while a Remote is kept around and
then calls list_references() on it to see if it still works. Apparently
it does, at least the address sanitizer does not complain.

Proposed-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

Finii commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ subprojects/*/
 docs/html/
 .vscode/settings.json
 .vscode/cspell.json
+unit_test_files
```
Why do we need this?

In my case the tests were conducted in `build/tests/unit_test_files/` which is already ignored because of the `build*/` rule.

And if, should it not be `unit_test_files/` instead?

alt-graph replied:
I happen to run the unit tests in the root of the working copy. But if we really want to debate every entry in .gitignore, it might be a better solution to remove it from the repo entirely and let everyone have their own copy. I'll add a slash for now.

Finii replied:
> But if we really want to debate every entry in .gitignore, it might be a better solution to remove it

Well, obviously the entry was not what has been intended. You want to ignore the directory even in the root dir (answering my question what it is needed for), but what you did was ignoring the directory and all also files with that name.

Is a review not to point out where specified / written-out things do not match intended things?

And yes, I think every entry in `.gitignore` should be open to discussion as every other line added to any other file.

Writing it without a slash implies some other use.
* I ask what your intention is
* I report what my usecase would be
* I suggest a fix

:woman_shrugging:

alt-graph committed:
.gitignore: Add a slash to the unit_test_files entry

Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph commented:
@Finii: I'll try to break your last comment apart to have threads that we can work on.

alt-graph commented:
> At least it should be called get_all_remotes() or similar.

I can live with `get_all_remotes()` as well, but I generally prefer using the most specific verb available instead of "get" if possible. So I still find `list_remotes()` quite descriptive, more compact, and therefore a tiny bit better than `get_all_remotes()`. Does anyone else have an opinion on that?

alt-graph commented:
>> In that case we do not need the Remote class at all
> Yes. The need for that class is indeed a question.

I see it as a way of making the access to remote-repo related settings easier. In any case I prefer not to have to deal with the `libgit2` C API directly.

> Anyhow, a Remote object by itself is not too helpful as it does not even remember which repo it belonged to. If I want to handle multiple repos and their remotes that becomes a hurdle. I as user need to keep the association myself somewhere. I would not need to do that if the remote is an integral part of the repository object.

If the "back-link" to the parent repo is really important to you, we can add a member function for that. I just saw no need for it so far because we always have the original repo around anyway?

> And then, why do we have the members AT ALL?
```cpp
class Remote
{
    /// Return a non-owning pointer to the underlying git remote object.
    git_remote* get() const { return remote_.get(); }

    std::string get_name() const { return git_remote_name(remote_.get()); }
    std::string get_url() const { return git_remote_url(remote_.get()); }

private:
    LibGitRemote remote_{ nullptr, git_remote_free };
};
```
That is fair enough, but it will crash if `git_remote_name()`/`git_remote_url()` return a null pointer, as they can according to the libgit2 docs. I would actually prefer this version over the current "caching" scheme. I still find it better than throwing the bare libgit2 C API at folks.

Finii commented:
_Unfortunately we are now stuck with a thread-less conversation that can not be resolved?_

> > At least it should be called get_all_remotes() or similar.
> 
> I can live with `get_all_remotes()` as well, but I generally prefer using the most specific verb available instead of "get" if possible. So I still find `list_remotes()` quite descriptive, more compact, and therefore a tiny bit better than `get_all_remotes()`. Does anyone else have an opinion on that?

I just think
* There might be users that just want a list of names, maybe to show in their GUI or something. Of course they could use the vector of `Remote`s and transform it into a vector of strings. The question is what is the more common use-case? And/or should we not support both? I can see use-cases for the string vector but I fail to see any for a `Remote` vector. Typically users would filter that again to find that one remote they are interested in?
* Regarding the name, you are right `get` is used far too much. What about `fetch_remotes`? `list` is in my eyes a more basic operation, and IF we have two functions (one for strings and one for objects), `list` should probably be for the strings?
* Maybe I just fail to see how you intend to use `list_remotes()`. I can hardly see a use-case for a string list. What would someone do with `Remote` objects to all configured remotes? Except filtering out ONE of them by name - which is in fact just `get_remote()`

Finii commented:
> I see it as a way of making the access to remote-repo related settings easier. In any case I prefer not to have to deal with the `libgit2` C API directly.

:+1:

> I just saw no need for it so far because we always have the original repo around anyway?

Well, usually 'is around' will prose problems later on.
We can leave it as it is and add a backlink later on.
 
> That is fair enough, but it will crash if `git_remote_name()`/`git_remote_url()` return a null pointer

Well, that might be prevented (like in the `safe_string` case above).
And also we would need to include the case where `remote_` is in fact `nullptr`.

> prefer this version over the current "caching" scheme.

After some thought maybe we should return `string_view` instead for cases where no string is needed, e.g. comparisons with a string literal.

> I still find it better than throwing the bare libgit2 C API at folks.

:+1:

Finii commented on tests/test_Remote.cc:
```diff
@@ -0,0 +1,110 @@
+TEST_CASE("Remote: list_references()", "[GitRepository]")
+{
+    const auto working_dir = unit_test_folder() / "Remote_list_references";
+    const auto remote_repo = unit_test_folder() / "Remote_list_references.remote";
```
After playing with this and the tests I wound rather call it

```suggestion
    const auto upstream_repo = unit_test_folder() / "Remote_list_references.upstream";
```

But then, maybe that is also wrong. Too many `remote` words.
"`upstream` is a remote of `repo`" is a bit easier ro grasp by me ;)

But then, this is very personal and maybe wrong.

alt-graph replied:
I think in strict git terminology a *remote branch* only becomes an *upstream* if it is set as the default target for pushing and pulling for some local branch. So that one can simply run `git push` or `git pull` without further arguments, I mean. This is not the case here.

Finii commented on tests/test_Remote.cc:
```diff
@@ -0,0 +1,110 @@
+    repo->commit("Add test.txt");
+
+    // Create a bare remote repository
+    repository_init(remote_repo, true);
```
First I thought we should not discard the return value here but can check if the push to `upstream` worked later on, but then, these are `Remote` tests, and so maybe not the right point here.

Finii commented on tests/test_Remote.cc:
```diff
@@ -0,0 +1,110 @@
+{
+    const auto reporoot = unit_test_folder() / "Remote";
+
+    std::filesystem::remove_all(reporoot);
```
This, and also lines 72/73:

Not sure why this code is in here.
If the tests are started the whole `unit_test_folder` get purged anyhow.
If this removing is needed, that means we reuse the directory for two tests (or at least expect that).
I would rather keep all tests' repositories in place so that one can see all artifacts (git repos) after the tests run via `cd ...; git remote -v` etc pp.
So I would NOT remove them here beforehand because that might prevent us from detecting someone who reuses the directory for another test.

Maybe the code is from before
```
tests/test_main.cc:    std::filesystem::remove_all(unit_test_folder());
```
and we have the same approach (remove individually) with `test_GitRepository.cc`. :thinking:

alt-graph replied:
Good point, removing the subdirectories is not necessary. Changed accordingly.

Finii commented on tests/test_GitRepository.cc:
```diff
@@ -575,19 +578,110 @@ TEST_CASE("GitRepository add() with glob", "[GitWrapper]")
+    REQUIRE(maybe_remote.has_value());
+    REQUIRE(maybe_remote->get_name() == "origin");
+    REQUIRE(maybe_remote->get_url() == repo_url);
+    REQUIRE(maybe_remote->get() != nullptr);
```
Maybe do all these tests also on `remote` of line 594?
Maybe superfluous, but who knows if `add_remote` fills `Remote` properly only because `get_remote` does?

alt-graph replied:
Yep, done.

Finii reviewed (approved):
Some small further findings and comments.

Finished also the functional part (did write/modify the tests and examined the repos etc) and it seems to work ;-) Of course.

Almost good to merge - act on the comments ad lib.

Finii commented:
> IMO repeating the namespace qualifier just adds visual noise and makes the code harder to read.

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/7c692613-3a50-4061-b714-ee4ebbfc5b10)

The visual and typing clutter by prepending all files in `tests/` with `test_` is also considerable :grimacing: :grin: 

It's `include/*/Remote`, `src/Remote`, and `tests/Remote` ah, no wait, `tests/test_Remote` :disappointed: 
At least for people working on the command line this is a hassle.

Not that this has anything to do with this PR.

alt-graph commented:
> The visual and typing clutter by prepending all files in `tests/` with `test_` is also considerable 😬 😁
> 
> It's `include/*/Remote`, `src/Remote`, and `tests/Remote` ah, no wait, `tests/test_Remote` 😞 At least for people working on the command line this is a hassle.

I agree with that.

alt-graph committed:
Remote: Do not cache values for name and URL

[why]
It is not necessary to spend CPU cycles for the respective libgit2 calls
on construction of the object if the information is never needed.

Proposed-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph commented:
>     * There might be users that just want a list of names, maybe to show in their GUI or something. Of course they could use the vector of `Remote`s and transform it into a vector of strings. The question is what is the more common use-case? And/or should we not support both? I can see use-cases for the string vector but I fail to see any for a `Remote` vector. Typically users would filter that again to find that one remote they are interested in?

Well, I have two remotes on several repositories and I have always found the output of `git remote` to be lacking information. Knowing that there are two remotes called "origin" and "github" is not as useful as getting a list of names **and URLs** as done by `git remote -v`. I guess that is why I prefer to return a list of objects and not just strings – the names are simply not enough information about a remote.

>     * Regarding the name, you are right `get` is used far too much. What about `fetch_remotes`? 

I would rather avoid that. "fetch" has a well-defined meaning in git-land.

>`list` is in my eyes a more basic operation, and IF we have two functions (one for strings and one for objects), `list` should probably be for the strings?

I have a different take on that, I guess. IF we have two functions, I would prefer `list_remotes()` to return a list of `Remote` objects and `list_remote_names()` to return a list of remote names.

>     * Maybe I just fail to see how you intend to use `list_remotes()`. I can hardly see a use-case for a string list. What would someone do with `Remote` objects to all configured remotes? Except filtering out ONE of them by name - which is in fact just `get_remote()`

Show them to a user so they can select the one they want. And for that, knowledge of the URL is important.

alt-graph committed:
GitRepository: Enhance tests for add_remote()

Proposed-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
Do not remove unit test subfolders

[why]
We are already deleting the entire unit_test_files folder at the
beginning of the test suite, so there is no need for individual tests to
remove their subdirectories.

Proposed-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

alt-graph committed:
GitRepository: Add list_remote_names()

[why]
To give users a simpler way of retrieving a list of repository names
(instead of retrieving Repository objects via list_remotes() and then
querying each one with get_name()).

Requested-by: Ulf Fini Jastrow <ulf.fini.jastrow@desy.de>
Signed-off-by: Lars Froehlich <lars.froehlich@desy.de>

Finii commented:
Your last comment about `git remove -v` is also true.

Thanks for considering all my confuse suggestions.

I would crate a PR now that drops the `Git` from `GitRepository`, to change that asap, because that is a major break?
